### PR TITLE
JetBrains: Add Perforce support

### PR DESCRIPTION
**Pull request by vdavid**
- Fixes https://github.com/sourcegraph/sourcegraph/issues/42353
- Fixes https://github.com/sourcegraph/sourcegraph/issues/40452
- Fixes https://github.com/sourcegraph/sourcegraph/issues/43183
- Fixes https://github.com/sourcegraph/sourcegraph/issues/41031
- Adds nicer error messages in case of git errors
- Makes sure we use the Git version the user wants to use

(This is a draft PR because of the "caveats" listed in the "Perforce" section, and for the lack of testing.)

## Summary and a few words about git support
This monster of a PR switches from using the `git` CLI command in the JetBrains plugin to using the **VCS feature built into JetBrains** IDEs.
Under the hood, JetBrains also uses the `git` CLI command, but the user can point to the file manually (who knows, the user might have several Git versions on their computer and want to choose which one to use for their project) and expect all parts of the IDE to use these settings. Before this PR, we used the CLI directly, which led some users to have problems with the plugin, and perhaps in other cases, we were using the wrong Git version all along without causing problems.

Although the `Git` plugin is bundled with all JetBrains products, it can be **disabled** or even uninstalled. As a result, in theory, not all users will have it installed. Until now, the plugin only worked with Git, so the Git plugin was a reasonable dependency. But now that we support Perforce, having the `Git` plugin shouldn't be required. So I've made it an **optional dependency**.

## Perforce support
Perforce works pretty differently from Git, as in, it's centralized, it doesn't have support for branching in the sense Git has, and the logic and terminology are different from Git.

My intention in this PR was to retain all functionality for Git and get get Perforce support on par with our Git support. For the most part, I've succeeded.

Caveats:
 - We currently don't support opening revision diffs for Perforce. I've created a low-priority issue to track this: https://github.com/sourcegraph/sourcegraph/issues/43805
 - Our `editor.go` back-end is not prepared for Perforce. I didn't want our customers to wait for a back-end change to roll out before they could use this feature, so in the case of Perforce, I skipped the `editor.go` redirect, and I'm going directly to our "file" and "search" pages. This has the benefit of avoiding the extra redirect (the page should load faster). On the other hand, we have a fairly elaborate `ReposourceCloneURLToRepoName` login in `editor.go` that we can't make use of this way. I've created an issue to track this: https://github.com/sourcegraph/sourcegraph/issues/43811
 - We only support custom URLs in a weird way. On our admin interface, we can specify a `repositoryPathPattern` ([example](https://cse-k8s.sgdev.org/site-admin/external-services/RXh0ZXJuYWxTZXJ2aWNlOjEyOQ==)) which modifies what the generated URLs should look like. Our "Remote URL replacements" feature should cover all cases, but it's non-intuitive. We should at least document this.

## Support for other VCSs
- I'm unsure if we ever had **Mercurial** support, but if we had, this change isn't likely to break it. As far as I know, Mercurial is pretty similar to Git, so it _might_ work out of the box. I only made some exceptions to recognize Perforce. In other cases, we fall back to the previous functionality. That said, I assume that the plugin **won't work** with Mercurial (not before or after this PR), and as far as I know, we don't support Mercurial overall, so this is a non-issue for now.
- We currently don't support either **SVN** or **CVS**, but with this switch to relying on the abstractions added by JetBrains, it should be a **lot** easier to add support for these in the future. (I think, as a company, we generally want to support these VCSs eventually.)

Perforce docs that helped, for the record:
 - [Docs: Using Perforce depots with Sourcegraph](https://docs.sourcegraph.com/admin/repo/perforce)
 - [Handbook: Dogfood Perforce server](https://handbook.sourcegraph.com/departments/ce-support/support/process/p4-enablement/)
 - [JetBrains: Configure Perforce integration](https://www.jetbrains.com/help/idea/enabling-and-configuring-perforce-integration.html)

## Test plan

[13-min Loom demo](https://www.loom.com/share/cf42ce45a2e94392958154e63311ae8f) where I go through all of these (and sorry for the two mess-ups in my demo 😂):
- Show that our "search" and open/copy" actions still work for Git repos
- Show that our "search" and open/copy" actions now work for Perforce repos (for https://github.com/sourcegraph/sourcegraph/issues/42353)
- Show that now we support Git even if the "git" command is not available in the CLI (for https://github.com/sourcegraph/sourcegraph/issues/40452)
- Show that "Open Revision Diff" now works for projects with multiple git repos (for https://github.com/sourcegraph/sourcegraph/issues/43183)
- Show that "Open Revision Diff" now doesn't error for Perforce repos (for https://github.com/sourcegraph/sourcegraph/issues/41031)
- Show that a git repo without an "origin" now throws much nicer errors

## App preview:

- [Web](https://sg-web-dv-jetbrains-perforce.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (0.00 kb) | 0.00% (0.00 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/dv/jetbrains-perforce/compare-report.html;tab=live_object) for a full comparison between the commits d6ffa71 and f1fb3aa or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/jetbrains/src/main/java/com/sourcegraph/vcs/GitUtil.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class GitUtil {
+
+    @NotNull
+    // Returned format: git@github.com:sourcegraph/sourcegraph.git
+    public static String getRemoteRepoUrl(@NotNull GitRepository repository, @NotNull Project project) throws Exception {
+        GitRemote remote = getBestRemote(repository, project);
```
Comment (philipp-spiess) [inline client/jetbrains/src/main/java/com/sourcegraph/vcs/GitUtil.java:20]: It's really nice that all of this is properly typed now!

File: client/jetbrains/src/main/java/com/sourcegraph/vcs/RepoUtil.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+import com.intellij.dvcs.repo.Repository;
+import com.intellij.dvcs.repo.VcsRepositoryManager;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.vcsUtil.VcsUtil;
+import com.sourcegraph.common.ErrorNotification;
+import com.sourcegraph.config.ConfigUtil;
+import git4idea.repo.GitRepository;
```
Comment (philipp-spiess) [inline client/jetbrains/src/main/java/com/sourcegraph/vcs/RepoUtil.java:11]: How does this work if the git integration is uninstalled (as mentioned in your PR description)?  Would this file throw when it tires to load? 🤔

File: client/jetbrains/src/main/java/com/sourcegraph/website/FileActionBase.java [hunk 3]
```diff
@@ -34,17 +36,30 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
         if (currentFile instanceof SourcegraphVirtualFile) {
             SourcegraphVirtualFile sourcegraphFile = (SourcegraphVirtualFile) currentFile;
-            handleFileUri(project, URLBuilder.buildSourcegraphBlobUrl(project, sourcegraphFile.getRepoUrl(), sourcegraphFile.getCommit(), sourcegraphFile.getRelativePath(), getSelectionStartPosition(editor), getSelectionEndPosition(editor)));
+            handleFileUri(project, URLBuilder.buildSourcegraphBlobUrl(project, sourcegraphFile.getRepoUrl(), sourcegraphFile.getCommit(), sourcegraphFile.getRelativePath(), selectionStartPosition, selectionEndPosition));
         } else {
-            RepoInfo repoInfo = GitUtil.getRepoInfo(currentFile.getPath(), project);
-            if (repoInfo.remoteUrl.equals("")) {
-                return;
-            }
+            ApplicationManager.getApplication().executeOnPooledThread(
```
Comment (philipp-spiess) [inline client/jetbrains/src/main/java/com/sourcegraph/website/FileActionBase.java:46]: Why is this necessary now? Maybe add a comment?

File: client/jetbrains/src/main/java/com/sourcegraph/website/OpenRevisionAction.java [hunk 4]
```diff
@@ -26,24 +29,95 @@
...
         }
-        if (revisionObject == null) {
+
+        // This action handles events for both log and history views, so attempt to load from any possible option.
+        RevisionContext context = getHistoryRevisionContext(event).or(() -> getLogRevisionContext(event)).orElse(null);
+
+        if (context == null) {
+            VirtualFile file = event.getDataContext().getData(VcsDataKeys.VCS_VIRTUAL_FILE);
+            if (file != null) {
+                // This cannot run on EDT
```
Comment (philipp-spiess) [inline client/jetbrains/src/main/java/com/sourcegraph/website/OpenRevisionAction.java:45]: What does `EDT` refer to?

**Review by philipp-spiess — APPROVED** [review]
There's a lot of stuff going on here and I’m sorry if I missed something underneath all the Java indirection 🙈 But this LGTM overall. Everything seems straight-forward. This PR is basically a lot of string manipulation 🙈

File: client/jetbrains/src/main/java/com/sourcegraph/website/OpenRevisionAction.java (hunk 4 above)
Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/website/OpenRevisionAction.java:45]: Event Dispatch Thread. I've added the resolution of this abbreviation everywhere.

File: client/jetbrains/src/main/java/com/sourcegraph/website/FileActionBase.java (hunk 3 above)
Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/website/FileActionBase.java:46]: I've now added reasoning everywhere (about 10 locations) where we do thread switches. Thanks!

File: client/jetbrains/src/main/java/com/sourcegraph/vcs/RepoUtil.java (hunk 2 above)
Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/vcs/RepoUtil.java:11]: Great question! I've tested it, it works; I think the IDE loads optional dependencies dynamically under the hood.
However, I was seeing some error messages in the console, so to make it slightly nicer for the user, I've added a proper check for the class existence, and in case of failure, I'm showing a helpful error message now.

**Comment by vdavid** [comment]
Thanks for the review, @philipp-spiess! I fixed up the threading-related comments and added a message in case of disabled Git/Perforce plugins, and added a bunch of docs to help users make this work with custom Perforce URLs. 

Also, I'll actually bump the version number to prepare for a release.
